src/number_theory/prime_counting.lean

```diff
@@ -0,0 +1,134 @@
+/-
+Copyright (c) 2021 Bolton Bailey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bolton Bailey
+-/
+
+import data.nat.prime
+import data.nat.totient
+import algebra.periodic
+import data.finset.locally_finite
+import data.nat.count
+
+
+/-!
+# The Prime Counting Function
+
+In this file we define the prime counting function - the function on natural numbers that returns
+the number of primes less than or equal to its input.
+
+## Main Results
+
+The main definitions for this file are
+
+- `prime_counting`: The prime counting function π
+- `prime_counting'`: π(n - 1)
+
+We then prove that these are monotone. The last main theorem is an upper bound on `π'` which arises
+by observing that all numbers greater than `k` and not coprime to `k` are not prime, and so only at
+most `φ(k)/k` fraction of the numbers from `k` to `n` are prime.
+
+## Notation
+
+We use the standard notation `π` to represent the prime counting function (and `π'` to represent
+the reindexed version).
+
+-/
+
+namespace nat
+open finset
+
+/--
+A variant of the traditional prime counting function which gives the number of primes
+*strictly* less than the input. More convenient for avoiding off-by-one errors.
+-/
+def prime_counting' : ℕ → ℕ := nat.count prime
+
+/-- The prime counting function: Returns the number of primes less than or equal to the input. -/
+def prime_counting (n : ℕ) : ℕ := prime_counting' (n + 1)
+
+localized "notation `π` := nat.prime_counting" in nat
+localized "notation `π'` := nat.prime_counting'" in nat
+
+lemma monotone_prime_counting' : monotone prime_counting' := count_monotone prime
+
+lemma monotone_prime_counting : monotone prime_counting :=
+λ a b a_le_b, monotone_prime_counting' (add_le_add_right a_le_b 1)
+
+private lemma filter_coprime_bound (a k n : ℕ) (a_pos : 0 < a) :
+  ((Ico k (k + n)).filter (coprime a)).card ≤ totient a * (n / a + 1) :=
+begin
+  conv
+  begin
+    to_lhs,
+    rw ←nat.mod_add_div n a,
+  end,
+  induction n / a with i ih,
+  { simp only [zero_add, mul_one, mul_zero, le_of_lt (mod_lt n a_pos)],
+    transitivity (filter a.coprime (Ico k (k + a))).card,
+    { mono,
+      refine monotone_filter_left a.coprime _,
+      simp only [add_zero, finset.le_eq_subset],
+      rw finset.subset_iff,
+      intro x,
+      simp only [and_imp, mem_Ico],
+      intros h1 h2,
+      refine ⟨h1, _⟩,
+      linarith only [h2, mod_lt n a_pos], },
+    { rw filter_coprime_Ico_eq_totient, }, },
```

Comment:
Are there useful sublemmas that can be extracted here? I would hope this base case of the induction can be shorter by have the appropriate lemmas.

Reply:
`Ico_subset_Ico` was what I was looking for, I just missed it with `library_search` since I was searching with the first argument to `Ico` equal.

Reply:
There should also be a lemma `(a le b) -> (b le c) -> Ico a c \subseteq Ico a b \cup Ico b c` though. I'll look into that.

Reply:
Ok, I've made a separate PR for that.